wallet/handler.go

```diff
@@ -259,6 +266,75 @@ func DescribeKey(store Store, req *DescribeKeyRequest) (*DescribeKeyResponse, er
 	return resp, nil
 }
 
+type RotateKeyRequest struct {
+	Wallet            string
+	Passphrase        string
+	NewPublicKey      string
+	TXBlockHeight     uint32
+	TargetBlockHeight uint32
+}
+
+type RotateKeyResponse struct {
+	MasterPublicKey   string `json:"master_public_key"`
+	NewPublicKey      string `json:"new_public_key"`
+	Base64Transaction string `json:"base64_transaction"`
+}
+
+func RotateKey(store Store, req *RotateKeyRequest) (*RotateKeyResponse, error) {
+	w, err := getWallet(store, req.Wallet, req.Passphrase)
+	if err != nil {
+		return nil, err
+	}
+
+	mKeyPair, err := w.GetMasterKeyPair()
+	if err != nil {
+		return nil, err
+	}
+
+	newKeyPair, err := w.DescribeKeyPair(req.NewPublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get new key pair: %w", err)
+	}
+
+	inputData := commands.NewInputData(uint64(req.TXBlockHeight))
+	inputData.Command = &commandspb.InputData_KeyRotateSubmission{
+		KeyRotateSubmission: &commandspb.KeyRotateSubmission{
+			KeyNumber:     uint64(newKeyPair.Index()),
+			TargetBlock:   uint64(req.TargetBlockHeight),
+			Time:          0, // @TODO fill this
+			NewPubKeyHash: hex.EncodeToString(vgcrypto.Hash([]byte(req.NewPublicKey))),
+		},
+	}
+
+	data, err := proto.Marshal(inputData)
+	if err != nil {
+		return nil, fmt.Errorf("failed marshal key rotate submission input data: %w", err)
```

Comment:
super-nit-pick : I tried to uniformize error messages as such `couldn't do <something>`, where possible. As it's user facing, i would like to avoid inconsistencies.